hangups/conversation.py

```diff
@@ -805,75 +805,130 @@ def _add_conversation(self, conversation, events=[]):
+        """
+        conv = self._conv_dict.get(conv_id, None)
+        if conv is not None:
+            return conv
+
+        logger.debug('conversation %s not found', conv_id)
+        try:
+            res = yield from self._client.get_conversation(
+                hangouts_pb2.GetConversationRequest(
+                    request_header=self._client.get_request_header(),
+                    conversation_spec=hangouts_pb2.ConversationSpec(
+                        conversation_id=hangouts_pb2.ConversationId(
+                            id=conv_id)),
+                    include_event=False))
+        except exceptions.NetworkError as err:
+            logger.warning('failed to fetch unknown conversation %s: %s',
+                           conv_id, err)
+            return
+
+        return self._add_conversation(res.conversation_state.conversation)
 
     @asyncio.coroutine
     def _on_event(self, event_):
-        """Receive a hangouts_pb2.Event and fan out to Conversations."""
+        """Receive a hangouts_pb2.Event and fan out to Conversations.
+
+        Args:
+            event_: hangouts_pb2.Event instance
+        """
+        conv = yield from self._fetch_conversation(event_.conversation_id.id)
+        if conv is None:
+            return
         self._sync_timestamp = parsers.from_timestamp(event_.timestamp)
-        try:
-            conv = self._conv_dict[event_.conversation_id.id]
-        except KeyError:
-            logger.warning('Received Event for unknown conversation {}'
-                           .format(event_.conversation_id.id))
-        else:
-            conv_event = conv.add_event(event_)
-            # conv_event may be None if the event was a duplicate.
-            if conv_event is not None:
-                yield from self.on_event.fire(conv_event)
-                yield from conv.on_event.fire(conv_event)
+        conv_event = conv.add_event(event_)
+        # conv_event may be None if the event was a duplicate.
+        if conv_event is not None:
+            yield from self.on_event.fire(conv_event)
+            yield from conv.on_event.fire(conv_event)
 
+    @asyncio.coroutine
     def _handle_conversation(self, conversation):
-        """Receive Conversation and create or update the conversation."""
+        """Receive Conversation and create or update the conversation.
+
+        Args:
+            conversation: hangouts_pb2.Conversation instance
+        """
         conv_id = conversation.conversation_id.id
         conv = self._conv_dict.get(conv_id, None)
         if conv is not None:
             conv.update_conversation(conversation)
+
+        elif not (len(conversation.self_conversation_state
+                      .delivery_medium_option)
+                  and (conversation.self_conversation_state.self_read_state
+                       .latest_read_timestamp)):
+            # this is a delta, critical entrys are missing
+            yield from self._fetch_conversation(conv_id)
```

Comment:
I think we can assume the conversation from the state update is always a delta, so this conditional can be simplified to:
```python
        if conv is None:
            # Ignore the delta and fetch the complete conversation.
            yield from self._fetch_conversation(conv_id)
        else:
            # Update conversation using the delta.
            conv.update_conversation(conversation)
```